### ISSUE TITLE
Add pre/post config to ToC

### ISSUE THREAD

**Issue by uPagge** [gohugoio/hugo#8338]
According to the html5 standard, there must be a title in the nav tag. However, the TOC is generated without the ability to specify this header

It should be like this. This helps visually impaired browsers to read what this section is responsible for.

```html
<nav>
  <h1>Navigation</h1>
  <ol>
    <li></li>
    <li></li>
  </ol>
</nav>
```

P.S. If you wish, you can make sure that this is done in Google. They add an h tag for the sections while hiding it, since it is only needed for semantic markup.

![изображение](https://user-images.githubusercontent.com/40397740/111505222-7c2b8c80-8759-11eb-99f8-b0fd940d51b7.png)
![изображение](https://user-images.githubusercontent.com/40397740/111505264-851c5e00-8759-11eb-8f71-138aba1f2ced.png)

**Comment by uPagge** [comment]
In any case, reference to the standard

https://html.spec.whatwg.org/multipage/sections.html#the-nav-element

**Comment by davidsneighbour** [comment]
It' the first time I am reading that the standard requires a title tag. It doesn't, and h1 is NOT a title tag. It's the topmost header tag and the standard _suggests_ to have only one in one single document. That's a suggestion. Having an h1 in an automatically created TOC makes no sense for 99% of pages and adding an h1 title to a nav that isn't even the sites main navigation will result not only in SEO issues but issues in all parsers that try to build a structure from your page. 

Not sure where you got that from, the only reason I would expect some form of title, but then in form of an title-attribute, would be for a11y if the navigation is not the main navigation.

**Comment by uPagge** [comment]
Perhaps the problem is in my English, please forgive me in advance, as I have to use a translator. An error in translation, it does not require a title, but an h tag. This is optional for the nav, but it is good practice as screen readers will be able to read this section for people with vision problems.

Originally in the html5 spec there was no such thing as h1-h6, there was only an h tag. And the numbering remained only for backward compatibility with html4

If you look at the specification, you will see that more than one h1 tag can be used and the document is considered valid in html5.

![image](https://user-images.githubusercontent.com/40397740/111633610-bdc64100-8806-11eb-99cc-08d963ab66c1.png)

https://html.spec.whatwg.org/#the-h1,-h2,-h3,-h4,-h5,-and-h6-elements

Problems with parsers arise now, because the nav section does not have an h tag, instead of a heading there is "Untitled Section"

"Not sure where you got that from"

From html5 spec

**Comment by davidsneighbour** [comment]
Screenreader: `aria-title` attribute on the `nav` tag. Sorry, but what you want is really something that might have been thought about 10 years ago but is not part of the "living standard" of HTML. It's a specific use case for your project, I understand, but making it accessible means using aria-attributes on the tags, not adding more tags inside.

**Comment by jmooring** [comment]
Perhaps an approach where one could override the HTML that encapsulates the outer `<ul>...<\ul>`

Then you could do:

```
[markup]
  [markup.tableOfContents]
    pre = "<nav id="TableOfContents" aria-label="Main"><h1>Something</h1>"
    post = "</nav>"
```

There's precedent for the `pre` and `post` approach with [Menu Entry Variables](https://gohugo.io/variables/menus/#menu-entry-variables).

**Comment by uPagge** [comment]
@jmooring Oh, thanks a lot, I will definitely try this way

**Comment by jmooring** [comment]
This was a suggestion for a software enhancement. It is not an active feature.

**Comment by bep** [comment]
@jmooring good suggestion. We do also have an issue about exposing the ToC as a data structure, allowing people to have their own templates, but I guess this is a easy improvement.

**Comment by nternetinspired** [comment]
I'd suggest that `aria-label` may be more appropriate in this situation, given that it's addition will not visually affect existing sites. In additon, I would suggest that an ordered list, `<ol>` is a better choice for a table of contents as the order in which the items are presented _does_ matter, the content author explicitly defined that order and presenting the items in a different order (as you can with an unordered list) would affect the meaning.

:100: should not be an `<h1>`!

**Comment by uPagge** [comment]
@nternetinspired 

```yaml
markup:
  tableOfContents:
    ordered: true
```

**Comment by nternetinspired** [comment]
…yes, but I'd argue that the default should semantically correct.

**Comment by jmooring** [comment]
...yes, but I'd argue that changing the default would break a lot of sites. Those who want ordered list have already configured their sites accordingly.

**Comment by nternetinspired** [comment]
It would _fix_ a lot of sites. Using an unordered list for a table of contents is incorrect. Changing that may lead to some _minor_ visual changes but document structure and accessibility would be improved.

Just my $0.02 and happy to leave it at that. As already pointed out, I can correct the semantics of my own sites quite easily, but I beleive it would be preferable for those with less experience to not have to. My general preferance will always be for ‘better by default’.

**Comment by davidsneighbour** [comment]
Yep, let's leave it at that. There is no such thing as a TOC in HTML. Using a list for an overview of headings is not intended. Be it unordered or ordered. In the current HTML standard the nearest that would come to something like a TOC would be a nav element with subsequent links to the sections of the page. If one of these links would have child-sections then some form of sub-division is required. which could be a div or a list. 

Having the TOC as data structure will solve all these issues and keeps Hugo out of religious conversations (like this one).

**Comment by jmooring** [comment]
@davidsneighbour,

I respect the ideas and concerns raised by both @uPagge and @nternetinspired. The title of this issue is  "Add pre/post config to ToC" and it has been flagged as a "GoodFirstIssue." If we are going to take a swing at incremental improvement before implementing [TOC as a data structure](https://github.com/gohugoio/hugo/issues/5087) (which has been open for 31 months) then conversations like this are important and healthy.

For the record, I agree with @nternetinspired. If you are going to use list items to display a table of contents, then the list items should be ordered. I just don't want to break existing sites without a really good reason. 

Out of curiosity I looked at a few higher-profile Hugo sites (digital.gov, getbootstrap.com, kubernetes.io), and they would all be visually impacted by a `<ul>` to `<ol>` TOC change. Of those three sites, digital.gov would see the most change.

Before

![image](https://user-images.githubusercontent.com/335669/111824309-48f51480-88a3-11eb-89e9-2dbe846a4d2f.png)

After

![image](https://user-images.githubusercontent.com/335669/111824471-7f329400-88a3-11eb-9d5c-572a794dfd31.png)

**Comment by uPagge** [comment]
In fact, I find it strange that as a developer I cannot influence this area and make it the way I want it.

It seems to me that adding a similar feature is an excellent solution to this problem.

```yaml
[markup]
  [markup.tableOfContents]
    pre = ""
    post = ""
```

By default, you can leave it as it is generated now, but give the opportunity to override this behavior.

**Comment by nternetinspired** [comment]
Apologies if my comments have ruffled feathers, that was never my intention. I am often direct and tone-of-voice does not easily transfer to type. I should have introduced myself!

Hi 👋 I'm Seth. I try to make user-centric sites, both professionally and in my free time. I love Hugo and I'd like to contribute where I can.

@jmooring Thanks. Agree, there would be a _significant_, not minor, change visually. My primary concern was a11y and UX.

**Comment by bep** [comment]
We're not changing the default.

**Comment by xbc5** [comment]
I was directed here for my issue (#10371), which is somewhat related.

In addition: if you don't want to change the default (for any reason) perhaps create a new behaviour, and eventually (possibly) deprecate the old one.

**Comment by Paul-Stern** [comment]
Hello everyone! My name is Paul, I'm a newcomer here and I would like to contribute to this issue. The issue is open but a quick look at the discussion above casts some doubts whether it still needs any work to be done. So is it supposed to make any changes to the project's code about this issue now? I am asking as I was looking for an issue for my first contributions and found this one.

**Comment by EPoY74** [comment]
Hello! Is this issues still relevant?

**Comment by dohrisalim** [comment]
I can implement pre/post content configuration for table of contents generation to meet HTML5 standards. I understand the need to add header elements and custom content around the TOC.

**Comment by manojponnusamy2032-star** [comment]
I would like to work on this issue. Is it still available?

**Comment by jmooring** [comment]
This proposal was approved a few years before the implementation of `Page.Fragments.ToHTML`. Since then we've seen how this method has been implemented in the wild... site and theme authors are adding site parameters for `startLevel`, `endLevel`, and `ordered` because (a) they need default values, and (b) they cannot read `Site.markup.tableOfContents` because that config is not exposed to the templates.

Presuming we add `pre` and `post` to the config struct, I think it would make sense to have `Page.Fragments.ToHTML` take an _optional_ options map instead of 3 to 5 args. For example:


```text
{{ $opts := dict
  "startLevel" 2
  "endLevel" 4
  "ordered" true
  "pre" "<nav>"
  "post" "</nav>"
}}
{{ .Fragments.ToHTML $opts }}
```

Any option not defined in the map would fall back to the site config (and its default values). This change would be backward compatible... if the first arg is a map (or there are zero args) do this new thing, else do the old thing.

And we should probably formally deprecate the old (3 args) thing.

Moving this back to proposal pending bep's review.
